File: src/remoteentity/RemoteEntityFormatJson.cpp
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.
//SOFTWARE.

#include "finalmq/remoteentity/RemoteEntityFormatJson.h"

#include "finalmq/serializejson/SerializerJson.h"
#include "finalmq/serializeproto/SerializerProto.h"
#include "finalmq/serializestruct/ParserStruct.h"
#include "finalmq/serializejson/ParserJson.h"
#include "finalmq/serializeproto/ParserProto.h"
#include "finalmq/serializestruct/SerializerStruct.h"
#include "finalmq/serializestruct/StructFactoryRegistry.h"
#include "finalmq/protocolsession/ProtocolMessage.h"
#include "finalmq/helpers/ZeroCopyBuffer.h"

#include "finalmq/remoteentity/entitydata.fmq.h"

//#include "finalmq/helpers/ModulenameFinalmq.h"


using finalmq::MsgMode;
using finalmq::Header;


namespace finalmq {

const int RemoteEntityFormatJson::CONTENT_TYPE = 2;
const std::string RemoteEntityFormatJson::CONTENT_TYPE_NAME = "json";

const std::string RemoteEntityFormatJson::PROPERTY_SERIALIZE_ENUM_AS_STRING = "enumAsSt";
const std::string RemoteEntityFormatJson::PROPERTY_SERIALIZE_SKIP_DEFAULT_VALUES = "skipDefVal";



struct RegisterFormatJson
{
    RegisterFormatJson()
    {
        RemoteEntityFormatRegistry::instance().registerFormat(RemoteEntityFormatJson::CONTENT_TYPE_NAME, RemoteEntityFormatJson::CONTENT_TYPE, std::make_shared<RemoteEntityFormatJson>());
    }
} g_registerFormatJson;



#define JSONBLOCKSIZE   512

void RemoteEntityFormatJson::serialize(const IProtocolSessionPtr& session, IMessage& message, const Header& header, const StructBase* structBase)
{
    message.addSendPayload("[", 1, JSONBLOCKSIZE);

    SerializerJson serializerHeader(message, JSONBLOCKSIZE);
    ParserStruct parserHeader(serializerHeader, header);
    parserHeader.parseStruct();


    // add end of header
    if (structBase)
    {
        // delimiter between header and payload
        message.addSendPayload(",\t", 2, JSONBLOCKSIZE);

        serializeData(session, message, structBase);

        message.addSendPayload("]\t\t", 3);
    }
    else
    {
        message.addSendPayload(",\t{}]\t\t", 7);
    }
}


void RemoteEntityFormatJson::serializeData(const IProtocolSessionPtr& session, IMessage& message, const StructBase* structBase)
{
    if (structBase)
    {
        bool enumAsString = true;
        bool skipDefaultValues = false;
        const Variant& formatData = session->getFormatData();
        if (formatData.getType() != VARTYPE_NONE)
        {
            const bool* propEnumAsString = formatData.getData<bool>(PROPERTY_SERIALIZE_ENUM_AS_STRING);
            const bool* propSkipDefaultValues = formatData.getData<bool>(PROPERTY_SERIALIZE_SKIP_DEFAULT_VALUES);
            if (propEnumAsString != nullptr)
            {
                enumAsString = *propEnumAsString;
            }
            if (propSkipDefaultValues != nullptr)
            {
                skipDefaultValues = *propSkipDefaultValues;
            }
        }

        // payload
        if (structBase->getRawContentType() == CONTENT_TYPE)
        {
            const std::string* rawData = structBase->getRawData();
            assert(rawData);
            char* payload = message.addSendPayload(rawData->size());
            memcpy(payload, rawData->data(), rawData->size());
        }
        else if (structBase->getStructInfo().getTypeName() == GeneralMessage::structInfo().getTypeName())
        {
            const GeneralMessage* generalMessage = static_cast<const GeneralMessage*>(structBase);
            if (!generalMessage->type.empty())
            {
                SerializerJson serializerData(message, JSONBLOCKSIZE, enumAsString, skipDefaultValues);
                ParserProto parserData(serializerData, generalMessage->data.data(), generalMessage->data.size());
                parserData.parseStruct(generalMessage->type);
            }
            else
            {
                message.addSendPayload("{}", 2);
            }
        }
        else if (structBase->getStructInfo().getTypeName() != RawDataMessage::structInfo().getTypeName())
        {
            SerializerJson serializerData(message, JSONBLOCKSIZE, enumAsString, skipDefaultValues);
            ParserStruct parserData(serializerData, *structBase);
            parserData.parseStruct();
        }
    }
    else
    {
        message.addSendPayload("{}", 2);
    }
}


static ssize_t findLast(const char* buffer, ssize_t size, char c)
{
    for (ssize_t i = size - 1; i >= 0; --i)
    {
        if (buffer[i] == c)
        {
            return i;
        }
    }
    return -1;
}



static size_t findEndOfPath(const char* buffer, ssize_t size)
{
    int i = 0;
    char cOld = 0;
    char c;
    while ((c = buffer[i]) && (i < size))
    {
        if (c == '{')
        {
            char cNext = buffer[i + 1];
            if (cOld != '/' && (cNext == '\"' || cNext == '}'))
            {
                return i;
            }
        }
        cOld = c;
        ++i;
    }
    return i;
}




static const std::string FMQ_PATH = "fmq_path";



std::shared_ptr<StructBase> RemoteEntityFormatJson::parse(const IProtocolSessionPtr& session, const BufferRef& bufferRef, bool storeRawData, const std::unordered_map<std::string, hybrid_ptr<IRemoteEntity>>& name2Entity, Header& header, int& formatStatus)
{
    formatStatus = 0;
    char* buffer = bufferRef.first;
    ssize_t sizeBuffer = bufferRef.second;

    if (sizeBuffer == 0)
    {
        return nullptr;
    }

    if (sizeBuffer > 0 && buffer[0] == '[')
    {
        ++buffer;
        --sizeBuffer;
    }
    if (sizeBuffer > 0 && buffer[sizeBuffer - 1] == ']')
    {
        --sizeBuffer;
    }

    std::string typeOfGeneralMessage;

    static const std::string WILDCARD = "*";
    const char* endHeader = nullptr;
    if (buffer[0] == '/')
    {
        // 012345678901234567890123456789
        // /MyServer/test.TestRequest!1{}
        size_t ixEndHeader = findEndOfPath(buffer, sizeBuffer);   //28
        endHeader = &buffer[ixEndHeader];

        ssize_t ixCorrelationId = findLast(buffer, ixEndHeader, '!');   //26
        if (ixCorrelationId != -1)
        {
            header.corrid = strtoll(&buffer[ixCorrelationId + 1], nullptr, 10);
        }
        else
        {
            ixCorrelationId = ixEndHeader;
        }

        std::string pathWithoutFirstSlash = { &buffer[1], &buffer[ixCorrelationId] };
        pathWithoutFirstSlash.erase(pathWithoutFirstSlash.find_last_not_of(" \n\r\t") + 1);
        size_t maxMatchLength = 0;
        const std::string* foundEntityName = nullptr;
        hybrid_ptr<IRemoteEntity> remoteEntity;
        for (auto it = name2Entity.begin(); it != name2Entity.end(); ++it)
        {
            const std::string& prefix = it->first;
            if (prefix != WILDCARD &&
                pathWithoutFirstSlash.size() >= prefix.size() &&
                pathWithoutFirstSlash.compare(0, prefix.size(), prefix) == 0 &&
                (pathWithoutFirstSlash.size() == prefix.size() || pathWithoutFirstSlash[prefix.size()] == '/') &&
                prefix.size() > maxMatchLength)
            {
                maxMatchLength = prefix.size();
                foundEntityName = &prefix;
                remoteEntity = it->second;
            }
        }
        if (foundEntityName)
        {
            header.destname = *foundEntityName;   // std::string(pathWithoutFirstSlash.c_str(), foundEntityName->size());
            if (pathWithoutFirstSlash.size() > foundEntityName->size())
            {
                header.path = std::string(&pathWithoutFirstSlash[foundEntityName->size() + 1], pathWithoutFirstSlash.size() - foundEntityName->size() - 1);
            }
        }
        else
        {
            header.destname = "*";
            header.path = pathWithoutFirstSlash;
            auto it = name2Entity.find(header.destname);
            if (it != name2Entity.end())
            {
                remoteEntity = it->second;
            }
        }
        auto entity = remoteEntity.lock();
        if (entity)
        {
            header.type = entity->getTypeOfCommandFunction(header.path, typeOfGeneralMessage);
        }

        std::string path = { &buffer[0], &buffer[ixCorrelationId] };
        path.erase(path.find_last_not_of(" \n\r\t") + 1);
        header.meta.emplace_back(FMQ_PATH);
        header.meta.emplace_back(std::move(path));

        header.mode = MsgMode::MSG_REQUEST;
    }
    else
    {
        SerializerStruct serializerHeader(header);
        ParserJson parserHeader(serializerHeader, buffer, sizeBuffer);
        endHeader = parserHeader.parseStruct(header.getStructInfo().getTypeName());
        if (endHeader)
        {
            // skip comma
            ++endHeader;
        }
        if (header.type.empty() && !header.path.empty())
        {
            hybrid_ptr<IRemoteEntity> remoteEntity;
            auto it = name2Entity.find(header.destname);
            if (it != name2Entity.end())
            {
                remoteEntity = it->second;
            }
            else
            {
                it = name2Entity.find(WILDCARD);
                if (it != name2Entity.end())
                {
                    remoteEntity = it->second;
                }
            }
            auto entity = remoteEntity.lock();
            if (entity)
            {
                header.type = entity->getTypeOfCommandFunction(header.path, typeOfGeneralMessage);
            }
        }
        if (header.path.empty() && !header.type.empty())
        {
            header.path = header.type;
        }
        else if (!header.path.empty() && header.type.empty())
        {
            header.type = header.path;
        }
    }

    std::shared_ptr<StructBase> data;

    if (endHeader)
    {
        ssize_t sizeHeader = endHeader - buffer;
        assert(sizeHeader >= 0);
        buffer += sizeHeader;
        ssize_t sizeData = sizeBuffer - sizeHeader;
        assert(sizeData >= 0);

        BufferRef bufferRefData = {buffer, sizeData};
        data = parseData(session, bufferRefData, storeRawData, header.type, formatStatus, typeOfGeneralMessage);
    }

    return data;
}



std::shared_ptr<StructBase> RemoteEntityFormatJson::parseData(const IProtocolSessionPtr& /*session*/, const BufferRef& bufferRef, bool storeRawData, std::string& type, int& formatStatus, const std::string& typeOfGeneralMessage)
{
    formatStatus = 0;
    const char* buffer = bufferRef.first;
    ssize_t sizeBuffer = bufferRef.second;

    std::shared_ptr<StructBase> data;

    ssize_t sizeData = sizeBuffer;
    assert(sizeData >= 0);
    if (!type.empty())
    {
        data = StructFactoryRegistry::instance().createStruct(type);

        if (data)
        {
            if (sizeData > 0)
            {
                if (type != GeneralMessage::structInfo().getTypeName() || typeOfGeneralMessage.empty())
                {
                    SerializerStruct serializerData(*data);
                    ParserJson parserData(serializerData, buffer, sizeData);
                    const char* endData = parserData.parseStruct(type);
                    if (!endData)
                    {
                        formatStatus |= FORMATSTATUS_SYNTAX_ERROR;
                        data = nullptr;
                    }
                }
                else
                {
                    ZeroCopyBuffer serializationBuffer;
                    SerializerProto serializerData(serializationBuffer);
                    ParserJson parserData(serializerData, buffer, sizeData);
                    const char* endData = parserData.parseStruct(typeOfGeneralMessage);
                    if (endData)
                    {
                        GeneralMessage* generalMessage = static_cast<finalmq::GeneralMessage*>(data.get());
                        generalMessage->type = typeOfGeneralMessage;
                        std::string serializedData = serializationBuffer.getData();
                        serializationBuffer.copyData(generalMessage->data);
                    }
                    else
                    {
                        formatStatus |= FORMATSTATUS_SYNTAX_ERROR;
                        data = nullptr;
                    }
                }
            }
            else
            {
                if (type == GeneralMessage::structInfo().getTypeName() && !typeOfGeneralMessage.empty())
                {
                    GeneralMessage* generalMessage = static_cast<finalmq::GeneralMessage*>(data.get());
                    generalMessage->type = typeOfGeneralMessage;
                }
            }
        }
    }

    if (storeRawData)
    {
        if (data == nullptr)
        {
            data = std::make_shared<RawDataMessage>();
        }
        data->setRawData(type, CONTENT_TYPE, buffer, sizeData);
    }

    return data;
}





}   // namespace finalmq

File: csharp/src/remoteentity/RemoteEntityFormatJson.cs
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

using System.Diagnostics;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Text;

namespace finalmq
{

    public class RemoteEntityFormatJson : IRemoteEntityFormat
    {
        static readonly int CONTENT_TYPE = 2;
        static readonly string CONTENT_TYPE_NAME = "json";

        public static readonly string PROPERTY_SERIALIZE_ENUM_AS_STRING = "enumAsSt";
        public static readonly string PROPERTY_SERIALIZE_SKIP_DEFAULT_VALUES = "skipDefVal";

        static readonly int JSONBLOCKSIZE = 512;
        static readonly byte[] ARRAY_BEGIN = { (byte)'[' };
        static readonly byte[] ARRAY_END = { (byte)']', (byte)'\t' };
        static readonly byte[] ARRAY_END_NO_DATA = { (byte)',', (byte)'\t', (byte)'{', (byte)'}', (byte)']', (byte)'\t' };
        static readonly byte[] DELIMITER_HEADER_PAYLOAD = { (byte)',', (byte)'\t' };
        static readonly byte[] EMPTY_PAYLOAD = { (byte)'{', (byte)'}' };

        static readonly string WILDCARD = "*";

        static readonly string FMQ_PATH = "fmq_path";


#pragma warning disable CA2255 // Attribut "ModuleInitializer" nicht in Bibliotheken verwenden
        [ModuleInitializer]
#pragma warning restore CA2255 // Attribut "ModuleInitializer" nicht in Bibliotheken verwenden
        internal static void RegisterFormat()
        {
            RemoteEntityFormatRegistry.Instance.RegisterFormat(CONTENT_TYPE_NAME, CONTENT_TYPE, new RemoteEntityFormatJson());
        }

        static int FindEndOfPath(byte[] buffer, int offset, int size)
        {
            int i = offset;
            int iEnd = offset + size;
            while (i < iEnd)
            {
                byte c = buffer[i];
                if (c == '{')
                {
                    return i;
                }
                ++i;
            }
            return i;
        }

        public StructBase? Parse(IProtocolSession session, BufferRef bufferRef, bool storeRawData, IDictionary<string, IRemoteEntity> name2Entity, Header header, out int formatStatus)
        {
            formatStatus = 0;
            byte[] buffer = bufferRef.Buffer;
            int offset = bufferRef.Offset;
            int sizeBuffer = bufferRef.Length;

            if (sizeBuffer == 0)
            {
                return null;
            }

            if (sizeBuffer > 0 && buffer[offset + 0] == '[')
            {
                ++offset;
                --sizeBuffer;
            }
            if (sizeBuffer > 0 && buffer[offset + sizeBuffer - 1] == ']')
            {
                --sizeBuffer;
            }

            int endHeader = -1;
            if (buffer[offset + 0] == '/')
            {
                // 012345678901234567890123456789
                // /MyServer/test.TestRequest!1{}
                int ixEndHeader = FindEndOfPath(buffer, offset, sizeBuffer);   //28
                endHeader = ixEndHeader;

                string strHeader = Encoding.UTF8.GetString(buffer, offset, ixEndHeader);

                int ixCorrelationId = strHeader.LastIndexOf('!');
                if (ixCorrelationId != -1)
                {
                    try
                    {
                        header.corrid = UInt64.Parse(strHeader.Substring(ixCorrelationId + 1));
                    }
                    catch (Exception)
                    {
                    }
                }
                else
                {
                    ixCorrelationId = ixEndHeader;
                }

                string pathWithoutFirstSlash = strHeader.Substring(1, ixCorrelationId - 1);
                pathWithoutFirstSlash = pathWithoutFirstSlash.Trim();
                string? foundEntityName = null;
                IRemoteEntity? remoteEntity = null;
                foreach (var entry in name2Entity)
                {
                    string prefix = entry.Key;
                    if (prefix != WILDCARD &&
                        pathWithoutFirstSlash.Length >= prefix.Length &&
                        pathWithoutFirstSlash.StartsWith(prefix) &&
                        (pathWithoutFirstSlash.Length == prefix.Length || pathWithoutFirstSlash[prefix.Length] == '/'))
                    {
                        foundEntityName = prefix;
                        remoteEntity = entry.Value;
                    }
                }
                if (foundEntityName != null)
                {
                    header.destname = foundEntityName;  // pathWithoutFirstSlash.Substring(0, foundEntityName.Length);
                    if (pathWithoutFirstSlash.Length > foundEntityName.Length)
                    {
                        header.path = pathWithoutFirstSlash.Substring(foundEntityName.Length + 1, pathWithoutFirstSlash.Length - foundEntityName.Length - 1);
                    }
                }
                else
                {
                    header.destname = "*";
                    header.path = pathWithoutFirstSlash;
                    name2Entity.TryGetValue(header.destname, out remoteEntity);
                }
                if (remoteEntity != null)
                {
                    string pathTemp = header.path;
                    header.type = remoteEntity.GetTypeOfCommandFunction(ref pathTemp);
                    header.path = pathTemp;
                }

                string path = strHeader.Substring(0, ixCorrelationId);
                path = path.Trim();
                header.meta.Add(FMQ_PATH);
                header.meta.Add(path);

                header.mode = MsgMode.MSG_REQUEST;
            }
            else
            {
                SerializerStruct serializerHeader = new SerializerStruct(header);
                ParserJson parserHeader = new ParserJson(serializerHeader, buffer, offset, sizeBuffer);
                endHeader = parserHeader.ParseStruct(header.GetType().FullName!);
                if (endHeader != -1)
                {
                    // skip comma
                    ++endHeader;
                }
                if (header.type.Length == 0 && header.path.Length != 0)
                {
                    name2Entity.TryGetValue(header.destname, out var remoteEntity);
                    if (remoteEntity == null)
                    {
                        name2Entity.TryGetValue(WILDCARD, out remoteEntity);
                    }
                    if (remoteEntity != null)
                    {
                        string pathTemp = header.path;
                        header.type = remoteEntity.GetTypeOfCommandFunction(ref pathTemp);
                        header.path = pathTemp;
                    }
                }
                if (header.path.Length == 0 && header.type.Length != 0)
                {
                    header.path = header.type;
                }
                else if (header.path.Length != 0 && header.type.Length == 0)
                {
                    header.type = header.path;
                }                
            }
            
            StructBase? data = null;

            if (endHeader != -1)
            {
                int sizeHeader = endHeader - offset;
                Debug.Assert(sizeHeader >= 0);
                offset += sizeHeader;
                int sizeData = sizeBuffer - sizeHeader;
                Debug.Assert(sizeData >= 0);

                BufferRef bufferRefData = new BufferRef( buffer, offset, sizeData );
                data = ParseData(session, bufferRefData, storeRawData, header.type, ref formatStatus);
            }

            return data;
        }
        public StructBase? ParseData(IProtocolSession session, BufferRef bufferRef, bool storeRawData, string type, ref int formatStatus)
        {
            byte[] buffer = bufferRef.Buffer;
            int offset = bufferRef.Offset;
            int sizeBuffer = bufferRef.Length;

            StructBase? data = null;

            if (type.Length != 0)
            {
                data = TypeRegistry.Instance.CreateStruct(type);
                Debug.Assert(sizeBuffer >= 0);

                if (data != null)
                {
                    if (sizeBuffer > 0)
                    {
                        SerializerStruct serializerData = new SerializerStruct(data);
                        ParserJson parserData = new ParserJson(serializerData, buffer, offset, sizeBuffer);
                        int endData = parserData.ParseStruct(type);
                        if (endData == -1)
                        {
                            formatStatus |= (int)FormatStatus.FORMATSTATUS_SYNTAX_ERROR;
                            data = null;
                        }
                    }
                }

                if (storeRawData)
                {
                    if (data == null)
                    {
                        data = new RawDataMessage();
                    }
                    data.SetRawData(type, CONTENT_TYPE, bufferRef);
                }
            }

            return data;
        }
        public void Serialize(IProtocolSession session, IMessage message, Header header, StructBase? structBase = null)
        {
            message.AddSendPayload(ARRAY_BEGIN, JSONBLOCKSIZE);

            SerializerJson serializerHeader = new SerializerJson(message, JSONBLOCKSIZE);
            ParserStruct parserHeader = new ParserStruct(serializerHeader, header);
            parserHeader.ParseStruct();

            // add end of header
            if (structBase != null)
            {
                // delimiter between header and payload
                message.AddSendPayload(DELIMITER_HEADER_PAYLOAD, JSONBLOCKSIZE);

                SerializeData(session, message, structBase);

                message.AddSendPayload(ARRAY_END, JSONBLOCKSIZE);
            }
            else
            {
                message.AddSendPayload(ARRAY_END_NO_DATA);
            }
        }
        public void SerializeData(IProtocolSession session, IMessage message, StructBase? structBase = null)
        {
            if (structBase != null)
            {
                // payload
                if (structBase.GetRawContentType() == CONTENT_TYPE)
                {
                    BufferRef? rawData = structBase.GetRawData();
                    Debug.Assert(rawData != null);
                    message.AddSendPayload(rawData);
                }
                else
                {
                    bool enumAsString = true;
                    bool skipDefaultValues = false;
                    Variant? formatData = session.FormatData;
                    if (formatData != null && formatData.VarType != Variant.VARTYPE_NONE)
                    {
                        Variant? propEnumAsString = formatData.GetVariant(PROPERTY_SERIALIZE_ENUM_AS_STRING);
                        Variant? propSkipDefaultValues = formatData.GetVariant(PROPERTY_SERIALIZE_SKIP_DEFAULT_VALUES);
                        if (propEnumAsString != null)
                        {
                            enumAsString = propEnumAsString;
                        }
                        if (propSkipDefaultValues != null)
                        {
                            skipDefaultValues = propSkipDefaultValues;
                        }
                    }

                    SerializerJson serializerData = new SerializerJson(message, JSONBLOCKSIZE, enumAsString, skipDefaultValues);
                    ParserStruct parserData = new ParserStruct(serializerData, structBase);
                    parserData.ParseStruct();
                }
            }
            else
            {
                message.AddSendPayload(EMPTY_PAYLOAD);
            }
        }
    }

}   // namespace finalmq
